storage/remote/write.go

```diff
@@ -97,19 +100,38 @@ func (rws *WriteStorage) ApplyConfig(conf *config.Config) error {
 		return err
 	}
 
-	hash := md5.Sum(append(cfgBytes, externalLabelBytes...))
-	if hash == rws.configHash {
+	configHash := md5.Sum(cfgBytes)
+	externalLabelHash := md5.Sum(externalLabelBytes)
+	externalLabelUnchanged := externalLabelHash == rws.externalLabelHash
+	if configHash == rws.configHash && externalLabelUnchanged {
 		level.Debug(rws.logger).Log("msg", "remote write config has not changed, no need to restart QueueManagers")
 		return nil
 	}
 
-	rws.configHash = hash
+	rws.configHash = configHash
+	rws.externalLabelHash = externalLabelHash
 
 	// Update write queues
 	newQueues := []*QueueManager{}
-	// TODO: we should only stop & recreate queues which have changes,
-	// as this can be quite disruptive.
+	newHashes := [][16]byte{}
+	newClientIndexes := []int{}
 	for i, rwConf := range conf.RemoteWriteConfigs {
+		b, err := json.Marshal(rwConf)
+		if err != nil {
+			return err
+		}
+
+		// Use RemoteWriteConfigs and its index to get hash. So if its index changed,
+		// the correspoinding queue should also be restarted.
+		hash := md5.Sum(b)
+		if i < len(rws.queues) && rws.hashes[i] == hash && externalLabelUnchanged {
```

Comment:
I am not familiar with remote-wirte code, but why should `conf.RemoteWriteConfigs[i]` be associated only with `rws.queues[i]`/`rws.hashes[i]`? Is it not possible that `hash` is at some different index in `rws.hashes`?

Reply:
@codesome Yes it's possible that the `hash` at some different index, then the corresponding queue will be restarted.

The queue will be reserved only when the config and its index both unchanged.

For example, config `c1, c2, c3, c4, c5` will have queue `q1, q2, q3, q4, q5`.

Then the config update to `c1, c2', c3, c5`, the config and index of `c1` and `c3` both unchanged, so the queue of them will be reserved. However, the config of `c2` has changed and the index of `c5` have changed, so their queue should be restarted. 

Finally the queue will be `q1, q2', q3, q5'`.

Reply:
Yeah, as Yao Zengzeng said right now we lock all of those together. It might be possible in the future to look across indices for hashes that were not changed, but right now our metrics use the index as a label. For example `prometheus_remote_storage_shards` will have a label value pair like `queue="0:http://cortex/api/prom/push"` so the index needs to be tied to a hash.